Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
+    return std::tuple { offset, attributes };
 }
 
-inline std::pair<PropertyTable::find_iterator, bool> WARN_UNUSED_RETURN PropertyTable::add(VM& vm, const ValueType& entry)
+inline std::tuple<PropertyOffset, unsigned, bool> WARN_UNUSED_RETURN PropertyTable::add(VM& vm, const ValueType& entry)
 {
-    ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(entry.offset));
+    ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(entry.offset()));
 
     // Look for a value with a matching key already in the array.
-    find_iterator iter = find(entry.key);
-    if (iter.first)
-        return std::make_pair(iter, false);
+    auto [entryIndex, index, offset, attributes] = find(entry.key());
+    if (offset != invalidOffset)
+        return std::tuple { offset, attributes, false };
 
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numAdds;
 #endif
 
     // Ref the key
-    entry.key->ref();
+    entry.key()->ref();
 
     // ensure capacity is available.
-    if (!canInsert()) {
-        rehash(vm, m_keyCount + 1);
-        iter = find(entry.key);
-        ASSERT(!iter.first);
+    if (!canInsert(entry)) {
+        rehash(vm, m_keyCount + 1, !canBeFitInCompact(entry));
+        std::tie(entryIndex, index, offset, attributes) = find(entry.key());
+        ASSERT_UNUSED(offset, offset == invalidOffset);
+        ASSERT_UNUSED(entryIndex, entryIndex == EmptyEntryIndex);
     }
 
     // Allocate a slot in the hashtable, and set the index to reference this.
-    unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    iter.first = &table()[entryIndex - 1];
-    *iter.first = entry;
+    entryIndex = usedCount() + 1;
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
+        auto* vector = bitwise_cast<uint8_t*>(indexVector & indexVectorMask);
+        vector[index] = entryIndex;
+        tableFromIndexVector(vector)[entryIndex - 1] = entry;
+    } else {
+        auto* vector = bitwise_cast<uint32_t*>(indexVector & indexVectorMask);
+        vector[index] = entryIndex;
+        tableFromIndexVector(vector)[entryIndex - 1] = entry;
+    }
 
     ++m_keyCount;
     
-    return std::make_pair(iter, true);
+    return std::tuple { entry.offset(), entry.attributes(), true };
 }
 
-inline void PropertyTable::remove(VM& vm, const find_iterator& iter)
+inline void PropertyTable::remove(VM& vm, KeyType key, unsigned entryIndex, unsigned index)
 {
-    // Removing a key that doesn't exist does nothing!
-    if (!iter.first)
-        return;
-
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numRemoves;
 #endif
 
     // Replace this one element with the deleted sentinel. Also clear out
     // the entry so we can iterate all the entries as needed.
-    m_index[iter.second] = deletedEntryIndex();
-    iter.first->key->deref();
-    iter.first->key = PROPERTY_MAP_DELETED_ENTRY_KEY;
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```

Comment:
If you like my `withIndexVector`, it can be used here too, since both sides of the branch are the same minus type of index vector